presto-main/src/main/java/com/facebook/presto/operator/DictionaryLookupJoinPageBuilder.java

```diff
@@ -153,8 +153,9 @@ private void appendProbeIndex(JoinProbe probe)
             return;
         }
         for (int index : probe.getOutputChannels()) {
-            // be aware that getRegionSizeInBytes could be expensive
-            probeBlockBytes += probe.getPage().getBlock(index).getRegionSizeInBytes(position, 1);
+            Block block = probe.getPage().getBlock(index);
+            // Estimate the size of the current row
+            probeBlockBytes += block.getSizeInBytes() / block.getPositionCount();
```

Comment:
Consider renaming `probeBlockBytes` too.

Reply:
(1) In general, `getRegionSizeInBytes` should be avoided anywhere outside `spi`. I have another PR #9514 to just remove the existing usage. Historically, we had some other regression using this interface. It would be nice to make it `protected` but that is not supported at the moment. So I would say let's do not calculate the row size at all for consistency.
(2) By "very off" you mean the row sizes are unevenly distributed and we just happen to use the average multiple times for a potentially huge/tiny row?

Reply:
(1) i understand
(2) exactly. And this potentially may cause another regression, especially for a huge row.

Reply:
Actually we do calculate the size only once before (which probably is wrong in estimating the logical size and needs revision):
```
// update memory usage
if (previousPosition == position) {
    return;
}
```

The regression comes from `DictionaryBlock::getRegionSizeInBytes` in our case. More precisely, new array:
![screen shot 2017-12-08 at 3 56 15 pm](https://user-images.githubusercontent.com/2192913/33789856-6976ce2e-dc30-11e7-870c-36cbd5886891.png)

Reply:
What about special-casing `length==1` in `DictionaryBlock#getRegionSizeInBytes`?

Reply:
@highker side question. What is the tool you made screenshot of for drilling down methods CPU utilization?

Reply:
@sopel39 it's `perf report` (http://www.brendangregg.com/perf.html)

Reply:
@findepi that will be too hacky right? It would be nice to avoid using `getRegionSizeInBytes` anywhere. We might in the future introduce another `Block` with expensive `getRegionSizeInBytes`

Reply:
@findepi, I'm now more leaning towards special-casing `length == 1` case now. We had some offline discussion about this and doesn't seem to have a simple solution to this. Regardless of the price we need to pay for `getRegionSizeInBytes`, the original estimation was already wrong if a dictionary block is nested deep in another block; so we may over count the real memory usage. Maybe we should specialize a new interface to get a single position that should be fast; or we wanna exam each `getRegionSizeInBytes` to specialize `length == 1` case.

Reply:
If you find this too hacky (I am ashamed, I don't.. yet), you can consider introducing `Block.getPositionSizeInBytes(int position)` (with default impl delegating to `getRegionSizeInBytes` and specialized impl in `DictionaryBlock`). What do you think?

@dain ?

Reply:
There you go; I'm thinking something similar @findepi

Reply:
@findepi, we had several discussion wrt how to calculate the region size. It turns out to be very complicated. For example, if a dictionary block is within an array or map block, `getRegionSizeInBytes` can be fed with an arbitrary range. So I updated the comments and showed that using average actually isn't that bad. The size should be bounded within a reasonable range. Hope that works for you.

Reply:
@findepi, the the problem that @highker points out where you have a dictionary block inside of an array block breaks the `length ==1` special casing (I had the same idea :( ).

I'm going to approve this for now as it stops the pain, but we can revisit if a better idea comes up.

Reply:
👍  @highker @dain I understand this scenario wouldn't work.

maybe that's obvious to you, but can you describe a bit more how do you come up with an array block based on a dictionary block? 
Is it some file reader producing this?

Reply:
@findepi yes, there are `LongDictionaryStreamReader` or `SliceDictionaryStreamReader`